src/modules/kubernetes.rs

```diff
@@ -144,55 +128,87 @@ pub fn module<'a>(context: &'a Context) -> Option<Module<'a>> {
 
     let kube_ctx = env::split_paths(&kube_cfg).find_map(get_kube_context)?;
 
-    let ctx_components: Vec<Option<KubeCtxComponents>> = env::split_paths(&kube_cfg)
-        .map(|filename| get_kube_ctx_component(filename, kube_ctx.clone()))
+    let ctx_components: Vec<KubeCtxComponents> = env::split_paths(&kube_cfg)
```

Comment:
The more I stare at this, the more this feels wrong: IMO this should end up in a `<Option<KubeCtxComponents>`  -> either the context is found in the contexts or not:

```yaml
current-context: work
contexts:
- context:
    cluster: work
    namespace: work-dev
    user: clusterUser
  name: 50hzk8s
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
```

i.e in `get_kube_ctx_component` we filter for the name, so there will at most be one KubeCtxComponents coming out of this, why should this become a vector...?

```rust
fn get_kube_ctx_component(filename: path::PathBuf, current_ctx: &str) -> Option<KubeCtxComponents> {
...
    let ctx_yaml = conf["contexts"].as_vec().and_then(|contexts| {
        contexts
            .iter()
            .filter_map(|ctx| Some((ctx, ctx["name"].as_str()?)))
            .find(|(_, name)| *name == current_ctx)  // <- HERE
    });

    let ctx_components = KubeCtxComponents {
...
    };

    Some(ctx_components)
}
```

And if it is None, the whole module should be aborted, IMO...

Reply:
E.g. this seems to have introduced the vec: https://github.com/starship/starship/pull/3569/files#diff-893760ffe78f91e7934d6da73675dbc83b63e62bbce7341f70310b0ab1187442L89

Reply:
> in get_kube_ctx_component we filter for the name, so there will at most be one KubeCtxComponents coming out of this, why should this become a vector...?

The only condition I can think of that would cause this is if there are multiple Kubernetes configs that define the same context. If that isn't just a hard error, then you need some way to resolve conflicting settings.

Is this allowed by Kubernetes at all? If so, what's the correct behavior?

Reply:
I think I understand now. The logic in using a vector of components is to implicitly (throughout the rest of the module) implement the merging logic detailed at https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#merging-kubeconfig-files

I'm not sure I'm a fan of having things spread out like this. Perhaps the move is to put the logic for implementing the merge procedure into `get_kube_ctx_components`? Right now, I think if you have incomplete context components e.g. one config that only defines a user, and another config that only defined the namespace, the module will incorrectly only pick up on the first one defined.

Also surprisingly, it seems we don't actually have a test case for config merging.

Reply:
I actually think it's now correct: 

> Never change the value or map key. Example: Preserve the context of the first file to set current-context. Example: If two files specify a red-user, use only values from the first file's red-user. Even if the second file has non-conflicting entries under red-user, discard them.

My reading is, that the last sentence implies that the first context found is used for everything, no matter if another config defines "non-conflicting" items (eg. your example: "one config that only defines a user, and another config that only defined the namespace").

Reply:
Huh, I guess that's right. (And we did get [a comment](https://github.com/starship/starship/pull/3569#issuecomment-1102130317) that it wasn't working properly, though I don't really know what the issue was).

Is there a way we can add a test for this?

Reply:
Ok, just verified this behaviour: if there is already a context with that name, only the config from that file is used and not "added" or overwritten from a second file: both ` one.yml` and `two.yml` contain a context with `name: work`, but the values only come from the context which was defined first:

```
λ  KUBECONFIG=$(pwd)/one.yml:$(pwd)/two.yml kubectl config view # no "namespace" in one.yml, but in two.yml
apiVersion: v1
clusters: null
contexts:
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: work-cluster1
    user: clusterUser1
  name: work
current-context: work
kind: Config
preferences: {}
users: null

λ  KUBECONFIG=$(pwd)/two.yml:$(pwd)/one.yml kubectl config view # no user in two.yml -> gets set to an empty string!
apiVersion: v1
clusters: null
contexts:
- context:
    cluster: docker-desktop
    user: docker-desktop
  name: docker-desktop
- context:
    cluster: work-cluster2
    namespace: work-dev2
    user: ""
  name: work
current-context: work
kind: Config
preferences: {}
users: null
```